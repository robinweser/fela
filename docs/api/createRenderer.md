# `createRenderer([config])`

Creates a Fela renderer which renders your selectors, keyframes, fonts and static styles.<br>
It caches all rendered styles to be able to reuse them on future rendering cycles.<br>

## Arguments
1. `config`(*Object?*): Optional renderer configuration. The most common use case adding [plugins](../advanced/Plugins.md) to process styles before they get cached. *(See [renderer configuration](Renderer.md#configuration) for further information)*.
2. `initialState`: Optional initial renderer state. Used for reuse a server-generated style-state. *(See [renderer getState()](Renderer.md#getState) for getting this state)*

## Returns
([Renderer](Renderer.md)): A Renderer instance.

## Example

```javascript
import { createRenderer } from 'fela'

const rule = props => ({
  backgroundColor: 'red',
  fontSize: props.size,
  color: 'blue'
})

const renderer = createRenderer()

renderer.render(rule, { size: '12px' }) // => c0 c0-dzm1d6


console.log(renderer.renderToString())
// .c0{background-color:red;color:blue}.c0-dzm1d6{font-size:12px}
```
