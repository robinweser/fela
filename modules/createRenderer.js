import generateContentHash from './utils/generateContentHash'
import sortedStringify from './utils/sortedStringify'
import getFontFormat from './utils/getFontFormat'

import cssifyKeyframe from './utils/cssifyKeyframe'
import cssifyObject from './utils/cssifyObject'

export default function createRenderer(config = { }, initialState = { }) {
  const renderer = {
    emitCounter: 0,
    emitOffset: typeof initialState.emitCounter === 'number' ? parseInt(initialState.emitCounter) : 0,
    listeners: [],
    keyframePrefixes: config.keyframePrefixes || [ '-webkit-', '-moz-' ],
    plugins: config.plugins || [ ],

    /**
     * clears the sheet's cache but keeps all listeners
     */
    clear() {
      renderer.fontFaces = ''
      renderer.keyframes = ''
      renderer.statics = ''
      renderer.rules = ''
      renderer.mediaRules = { }
      renderer.rendered = { }
      renderer.base = { }
      renderer.ids = [ ]

      // emit changes to notify subscribers
      renderer._emitChange()
    },

    /**
     * renders a new rule variation and caches the result
     *
     * @param {Function} rule - rule which gets rendered
     * @param {Object?} props - properties used to render
     * @return {string} className to reference the rendered rule
     */
    renderRule(rule, props = { }) {
      // rendering a rule for the first time
      // will create an ID reference
      if (renderer.ids.indexOf(rule) < 0) {
        renderer.ids.push(rule)

        // directly render the static base style to be able
        // to diff future dynamic style with those
        renderer.renderRule(rule, { })
      }

      // uses the reference ID and the props to generate an unique className
      const ruleId = renderer.ids.indexOf(rule)
      const className = 'c' + ruleId + renderer._generatePropsReference(props)

      // only if the cached rule has not already been rendered
      // with a specific set of properties it actually renders
      if (!renderer.rendered.hasOwnProperty(className)) {
        const style = renderer._processStyle(rule(props))
        renderer._renderStyle(className, style, renderer.base[ruleId])

        renderer.rendered[className] = renderer._didChange

        if (renderer._didChange) {
          renderer._didChange = false
          renderer._emitChange()
        }

        // keep static style to diff dynamic onces later on
        if (className === 'c' + ruleId) {
          renderer.base[ruleId] = style
        }
      }

      const baseClassName = 'c' + ruleId
      if (!renderer.rendered[className]) {
        return baseClassName
      }

      // returns either the base className or both the base and the dynamic part
      return className !== baseClassName ? baseClassName + ' ' + className : className
    },

    /**
     * renders a new keyframe variation and caches the result
     *
     * @param {Keyframe} keyframe - Keyframe which gets rendered
     * @param {Object?} props - properties used to render
     * @return {string} animationName to reference the rendered keyframe
     */
    renderKeyframe(keyframe, props = { }) {
      // rendering a Keyframe for the first time
      // will create cache entries and an ID reference
      if (renderer.ids.indexOf(keyframe) < 0) {
        renderer.ids.push(keyframe)
      }

      const propsReference = renderer._generatePropsReference(props)
      const animationName = 'k' + renderer.ids.indexOf(keyframe) + propsReference

      // only if the cached keyframe has not already been rendered
      // with a specific set of properties it actually renders
      if (!renderer.rendered.hasOwnProperty(animationName)) {
        const processedKeyframe = renderer._processStyle(keyframe(props))
        const css = cssifyKeyframe(processedKeyframe, animationName, renderer.keyframePrefixes)
        renderer.rendered[animationName] = true
        renderer.keyframes += css
        renderer._emitChange()
      }

      return animationName
    },

    /**
     * renders a new font-face and caches it
     *
     * @param {FontFace} fontFace - fontFace which gets rendered
     * @return {string} fontFamily reference
     */
    renderFont(family, files, properties = { }) {
      if (!renderer.rendered.hasOwnProperty(family)) {
        const fontFace = {
          fontFamily: '\'' + family + '\'',
          src: files.map(src => 'url(\'' + src + '\') format(\'' + getFontFormat(src) + '\')').join(',')
        }

        const fontProperties = [ 'fontVariant', 'fontWeight', 'fontStretch', 'fontStyle', 'unicodeRange' ]
        Object.keys(properties).filter(prop => fontProperties.indexOf(prop) > -1).forEach(fontProp => fontFace[fontProp] = properties[fontProp])

        const css = '@font-face{' + cssifyObject(fontFace) + '}'
        renderer.rendered[family] = true
        renderer.fontFaces += css
        renderer._emitChange()
      }

      return family
    },

    /**
     * renders static style and caches them
     *
     * @param {string|Object} style - static style to be rendered
     * @param {string?} selector - selector used to render the styles
     * @return {string} rendered CSS output
     */
    renderStatic(style, selector) {
      const reference = typeof style === 'string' ? style : selector

      if (!renderer.rendered.hasOwnProperty(reference)) {
        if (typeof style === 'string') {
          // remove new lines from template strings
          renderer.statics += style.replace(/\s{2,}/g, '')
        } else {
          renderer.statics += selector + '{' + cssifyObject(renderer._processStyle(style)) + '}'
        }

        renderer.rendered[reference] = true
        renderer._emitChange()
      }
    },

    /**
     * renders all cached styles into a single valid CSS string
     * clusters media query styles into groups to reduce output size

     * @return single concatenated CSS string
     */
    renderToString() {
      let css = renderer.fontFaces + renderer.statics + renderer.rules

      for (let media in renderer.mediaRules) {
        css += '@media ' + media + '{' + renderer.mediaRules[media] + '}'
      }

      return css + renderer.keyframes
    },

    /**
     * Adds a new subscription to get notified on every rerender
     *
     * @param {Function} callback - callback function which will be executed
     * @return {Object} equivalent unsubscribe method
     */
    subscribe(callback) {
      renderer.listeners.push(callback)
      return {
        unsubscribe: () => renderer.listeners.splice(renderer.listeners.indexOf(callback), 1)
      }
    },

    /**
     * Returns the current state of the renderer
     * Can be used to transport the renderer state from server to client
     *
     * @return {Object} Containing the renderer state
     */
    getState() {
      return {
        emitCounter: renderer.emitCounter
      }
    },

    /**
     * calls each listener with the current CSS markup of all caches
     * gets only called if the markup actually changes
     * gets aborted when styles are already rendered on the server
     *
     * @param {Function} callback - callback function which will be executed
     * @return {Object} equivalent unsubscribe method
     */
    _emitChange() {
      renderer.emitCounter++
      if(renderer.emitOffset > 0){
        renderer.emitOffset--
        return
      }
      const css = renderer.renderToString()
      renderer.listeners.forEach(listener => listener(css))
    },

    /**
     * generates an unique reference id by content hashing props
     *
     * @param {Object} props - props that get hashed
     * @return {string} reference - unique props reference
     */
    _generatePropsReference(props) {
      return generateContentHash(sortedStringify(props))
    },

    /**
     * pipes a style object through a list of plugins
     *
     * @param {Object} style - style object to process
     * @return {Object} processed style
     */
    _processStyle(style) {
      return renderer.plugins.reduce((processedStyle, plugin) => plugin(processedStyle), style)
    },


    /**
     * iterates a style object and renders each rule to the cache
     *
     * @param {string} className - className reference to be rendered to
     * @param {Object} style - style object which is rendered
     * @param {Object`} base - base style subset for diffing
     */
    _renderStyle(className, style, base = { }, pseudo = '', media = '') {
      const ruleset = Object.keys(style).reduce((ruleset, property) => {
        const value = style[property]
        // recursive object iteration in order to render
        // pseudo class and media class declarations
        if (value instanceof Object && !Array.isArray(value)) {
          if (property.charAt(0) === ':') {
            renderer._renderStyle(className, value, base[property], pseudo + property, media)
          } else if (property.substr(0, 6) === '@media') {
            // combine media query rules with an `and`
            const query = property.slice(6).trim()
            const combinedMedia = media.length > 0 ? media + ' and ' + query : query
            renderer._renderStyle(className, value, base[property], pseudo, combinedMedia)
          }
        } else {
          // diff styles with the base styles to only extract dynamic styles
          if (value !== undefined && !base.hasOwnProperty(property) || base[property] !== value) {
            // remove concatenated string values including `undefined`
            if (typeof value === 'string' && value.indexOf('undefined') > -1) {
              return ruleset
            }
            ruleset[property] = value
          }
        }
        return ruleset
      }, { })

      // add styles to the cache
      if (Object.keys(ruleset).length > 0) {
        const css = '.' + className + pseudo + '{' + cssifyObject(ruleset) + '}'
        renderer._didChange = true

        if (media.length > 0) {
          if (!renderer.mediaRules.hasOwnProperty(media)) {
            renderer.mediaRules[media] = ''
          }

          renderer.mediaRules[media] += css
        } else {
          renderer.rules += css
        }
      }
    }
  }

  // initial setup
  renderer.keyframePrefixes.push('')
  renderer.clear()

  return renderer
}
