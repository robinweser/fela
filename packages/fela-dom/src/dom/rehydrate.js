/* @flow */
import arrayEach from 'fast-loops/lib/arrayEach'
import { RULE_TYPE, KEYFRAME_TYPE, getRuleScore } from 'fela-utils'

import rehydrateSupportRules from './rehydration/rehydrateSupportRules'
import rehydrateRules from './rehydration/rehydrateRules'

import render from './render'

import type { DOMRenderer } from '../../../../flowtypes/DOMRenderer'
import rehydrateKeyframes from './rehydration/rehydrateKeyframes'

const CLASSNAME_REGEX = /[.][a-z0-9_-]*/gi

// TODO: static, font
export default function rehydrate(
  renderer: DOMRenderer,
  targetDocument: any = document
): void {
  render(renderer, targetDocument)

  arrayEach(targetDocument.querySelectorAll('[data-fela-type]'), node => {
    const rehydrationAttribute =
      node.getAttribute('data-fela-rehydration') || -1
    const rehydrationIndex =
      renderer.uniqueRuleIdentifier || parseInt(rehydrationAttribute, 10)

    // skip rehydration if no rehydration index is set
    // this index is set to -1 if something blocks rehydration
    if (rehydrationIndex !== -1) {
      const type = node.getAttribute('data-fela-type') || ''
      const media = node.getAttribute('media') || ''
      const support = node.getAttribute('data-fela-support') || ''
      const css = node.textContent

      renderer.uniqueRuleIdentifier = rehydrationIndex

      const reference = type + media + support
      renderer.nodes[reference] = node

      if (type === RULE_TYPE) {
        if (support) {
          rehydrateSupportRules(css, media, renderer.cache)
        } else {
          rehydrateRules(css, media, '', renderer.cache)
        }

        // On Safari, style sheets with IE-specific media queries
        // can yield null for node.sheet
        // https://github.com/robinweser/fela/issues/431#issuecomment-423239591
        if (node.sheet && node.sheet.cssRules) {
          const nodeReference = media + support

          arrayEach(node.sheet.cssRules, (rule, index) => {
            const selectorText = rule.conditionText
              ? rule.cssRules[0].selectorText
              : rule.selectorText

            const score = getRuleScore(
              renderer.ruleOrder,
              selectorText.split(CLASSNAME_REGEX)[1]
            )

            if (score === 0) {
              renderer.scoreIndex[nodeReference] = index
            }

            rule.score = score
          })
        }
      } else if (type === KEYFRAME_TYPE) {
        rehydrateKeyframes(css, renderer.cache)
      }
    }
  })
}
