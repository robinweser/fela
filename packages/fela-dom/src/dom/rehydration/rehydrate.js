/* @flow */
import arrayEach from 'fast-loops/lib/arrayEach'
import { RULE_TYPE } from 'fela-utils'

import rehydrateSupportRules from './rehydrateSupportRules'
import rehydrateRules from './rehydrateRules'

import type { DOMRenderer } from '../../../../../flowtypes/DOMRenderer'

// rehydration (WIP)
// TODO: static, keyframe, font
export default function rehydrate(renderer: DOMRenderer): void {
  arrayEach(document.querySelectorAll('[data-fela-type]'), node => {
    const type = node.getAttribute('data-fela-type') || ''
    const media = node.getAttribute('media') || ''
    const support = node.getAttribute('data-fela-support')
    const css = node.textContent

    if (type === RULE_TYPE) {
      if (support) {
        rehydrateSupportRules(css, media, renderer.cache)
      } else {
        rehydrateRules(css, media, '', renderer.cache)
      }
    }
  })
}
