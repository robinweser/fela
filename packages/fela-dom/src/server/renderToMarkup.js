/* @flow */
import arrayReduce from 'fast-loops/lib/arrayReduce'
import objectReduce from 'fast-loops/lib/objectReduce'
import {
  clusterCache,
  cssifySupportRules,
  sheetMap,
  RULE_TYPE,
} from 'fela-utils'

import createStyleTagMarkup from './createStyleTagMarkup'
import getRehydrationIndex from './getRehydrationIndex'

import type { DOMRenderer } from '../../../../flowtypes/DOMRenderer'

export default function renderToMarkup(renderer: DOMRenderer): string {
  const cacheCluster = clusterCache(renderer.cache, renderer.ruleOrder)

  const rehydrationIndex = getRehydrationIndex(renderer)

  let styleMarkup = objectReduce(
    sheetMap,
    (markup, type, key) => {
      if (cacheCluster[key].length > 0) {
        markup += createStyleTagMarkup(
          cacheCluster[key],
          type,
          '',
          rehydrationIndex,
          false,
          renderer.styleNodeAttributes
        )
      }

      return markup
    },
    ''
  )

  const support = cssifySupportRules(cacheCluster.supportRules)

  if (support) {
    styleMarkup += createStyleTagMarkup(
      support,
      RULE_TYPE,
      '',
      rehydrationIndex,
      true,
      renderer.styleNodeAttributes
    )
  }

  const mediaKeys = Object.keys({
    ...cacheCluster.supportMediaRules,
    ...cacheCluster.mediaRules,
  }).sort(renderer.sortMediaQuery)

  return arrayReduce(
    mediaKeys,
    (markup, media) => {
      // basic media query rules
      if (
        cacheCluster.mediaRules[media] &&
        cacheCluster.mediaRules[media].length > 0
      ) {
        markup += createStyleTagMarkup(
          cacheCluster.mediaRules[media],
          RULE_TYPE,
          media,
          rehydrationIndex,
          false,
          renderer.styleNodeAttributes
        )
      }

      // support media rules
      if (cacheCluster.supportMediaRules[media]) {
        const mediaSupport = cssifySupportRules(
          cacheCluster.supportMediaRules[media]
        )

        if (mediaSupport.length > 0) {
          markup += createStyleTagMarkup(
            mediaSupport,
            RULE_TYPE,
            media,
            rehydrationIndex,
            true,
            renderer.styleNodeAttributes
          )
        }
      }

      return markup
    },
    styleMarkup
  )
}
