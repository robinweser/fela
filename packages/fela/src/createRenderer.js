import { cssifyDeclaration } from 'css-in-js-utils'
import { arrayEach, arrayReduce, arrayFilter } from 'fast-loops'
import isPlainObject from 'isobject'

import {
  generateDeclarationReference,
  generateCombinedMediaQuery,
  generateCSSSelector,
  isMediaQuery,
  isNestedSelector,
  isUndefinedValue,
  isSupport,
  normalizeNestedProperty,
  processStyleWithPlugins,
  STATIC_TYPE,
  RULE_TYPE,
  KEYFRAME_TYPE,
  FONT_TYPE,
  CLEAR_TYPE,
} from 'fela-utils'

import cssifyFontFace from './cssifyFontFace'
import cssifyKeyframe from './cssifyKeyframe'
import cssifyKeyframeRule from './cssifyKeyframeRule'
import cssifyStaticStyle from './cssifyStaticStyle'
import generateAnimationName from './generateAnimationName'
import generateClassName from './generateClassName'
import generateFontSource from './generateFontSource'
import generateStaticReference from './generateStaticReference'
import getFontLocals from './getFontLocals'
import isSafeClassName from './isSafeClassName'
import toCSSString from './toCSSString'
import validateSelectorPrefix from './validateSelectorPrefix'
import sortMediaQuery from './sortMediaQuery'
import extractCustomClassName from './extractCustomClassName'

export default function createRenderer(config = {}) {
  let renderer = {
    listeners: [],
    keyframePrefixes: config.keyframePrefixes || ['-webkit-', '-moz-'],
    plugins: config.plugins || [],
    sortMediaQuery:
      config.sortMediaQuery || sortMediaQuery(config.mediaQueryOrder),
    supportQueryOrder: config.supportQueryOrder || [],
    styleNodeAttributes: config.styleNodeAttributes || {},
    propertyPriority: config.propertyPriority || {},
    ruleOrder: [
      /^:link/,
      /^:visited/,
      /^:hover/,
      /^:focus-within/,
      /^:focus/,
      /^:active/,
    ],
    selectorPrefix: validateSelectorPrefix(config.selectorPrefix),
    specificityPrefix: config.specificityPrefix || '',
    filterClassName: config.filterClassName || isSafeClassName,
    devMode: config.devMode || false,

    uniqueRuleIdentifier: 0,
    uniqueKeyframeIdentifier: 0,

    nodes: {},
    scoreIndex: {},
    // use a flat cache object with pure string references
    // to achieve maximal lookup performance and memoization speed
    cache: {},

    getNextRuleIdentifier() {
      return ++renderer.uniqueRuleIdentifier
    },

    getNextKeyframeIdentifier() {
      return ++renderer.uniqueKeyframeIdentifier
    },

    renderRule(rule, props = {}) {
      return renderer._renderStyle(rule(props, renderer), props)
    },

    renderKeyframe(keyframe, props = {}) {
      const resolvedKeyframe = keyframe(props, renderer)
      const processedKeyframe = processStyleWithPlugins(
        renderer,
        resolvedKeyframe,
        KEYFRAME_TYPE,
        props
      )

      const keyframeReference = cssifyKeyframeRule(processedKeyframe)

      if (!renderer.cache.hasOwnProperty(keyframeReference)) {
        // use another unique identifier to ensure minimal css markup
        const animationName =
          renderer.selectorPrefix + renderer.generateAnimationName(props)

        const cssKeyframe = cssifyKeyframe(
          processedKeyframe,
          animationName,
          renderer.keyframePrefixes,
          keyframeReference
        )

        const change = {
          type: KEYFRAME_TYPE,
          keyframe: cssKeyframe,
          name: animationName,
        }

        renderer.cache[keyframeReference] = change
        renderer._emitChange(change)
      }

      return renderer.cache[keyframeReference].name
    },

    generateAnimationName() {
      return generateAnimationName(renderer.getNextKeyframeIdentifier())
    },

    renderFont(family, files, properties = {}) {
      const { localAlias, ...otherProperties } = properties

      const fontReference = family + JSON.stringify(properties)
      const fontLocals = getFontLocals(localAlias)

      if (!renderer.cache.hasOwnProperty(fontReference)) {
        const fontFamily = toCSSString(family)

        const fontFace = {
          ...otherProperties,
          src: generateFontSource(files, fontLocals),
          fontFamily,
        }

        const cssFontFace = cssifyFontFace(fontFace)

        const change = {
          type: FONT_TYPE,
          fontFace: cssFontFace,
          fontFamily,
        }

        renderer.cache[fontReference] = change
        renderer._emitChange(change)
      }

      return renderer.cache[fontReference].fontFamily
    },

    renderStatic(staticStyle, selector, props = {}) {
      const staticReference = generateStaticReference(staticStyle, selector)

      if (!renderer.cache.hasOwnProperty(staticReference)) {
        const cssDeclarations = cssifyStaticStyle(staticStyle, renderer, props)

        const change = {
          type: STATIC_TYPE,
          css: cssDeclarations,
          selector,
        }

        renderer.cache[staticReference] = change
        renderer._emitChange(change)
      }
    },

    subscribe(callback) {
      renderer.listeners.push(callback)

      return {
        unsubscribe: () =>
          renderer.listeners.splice(renderer.listeners.indexOf(callback), 1),
      }
    },

    clear() {
      renderer.uniqueRuleIdentifier = 0
      renderer.uniqueKeyframeIdentifier = 0
      renderer.cache = {}

      renderer._emitChange({
        type: CLEAR_TYPE,
      })
    },

    _renderStyle(style = {}, props = {}) {
      const customClassName = extractCustomClassName(style)
      const processedStyle = processStyleWithPlugins(
        renderer,
        style,
        RULE_TYPE,
        props,
        renderer.unoptimizedPlugins || renderer.plugins
      )

      return (
        customClassName +
        renderer._renderStyleToClassNames(processedStyle).substr(1)
      )
    },

    _renderStyleToClassNames(style, pseudo = '', media = '', support = '') {
      let classNames = ''

      const applyPlugin = (processed, plugin) => plugin(processed, renderer)

      for (const property in style) {
        const value = style[property]

        if (isPlainObject(value)) {
          if (isNestedSelector(property)) {
            classNames += renderer._renderStyleToClassNames(
              value,
              pseudo + normalizeNestedProperty(property),
              media,
              support
            )
          } else if (isMediaQuery(property)) {
            const combinedMediaQuery = generateCombinedMediaQuery(
              media,
              property.slice(6).trim()
            )
            classNames += renderer._renderStyleToClassNames(
              value,
              pseudo,
              combinedMediaQuery,
              support
            )
          } else if (isSupport(property)) {
            const combinedSupport = generateCombinedMediaQuery(
              support,
              property.slice(9).trim()
            )
            classNames += renderer._renderStyleToClassNames(
              value,
              pseudo,
              media,
              combinedSupport
            )
          } else {
            console.warn(`The object key "${property}" is not a valid nested key in Fela.
Maybe you forgot to add a plugin to resolve it?
Check http://fela.js.org/docs/basics/Rules.html#styleobject for more information.`)
          }
        } else {
          let declarationReference = generateDeclarationReference(
            property,
            value,
            pseudo,
            media,
            support
          )

          if (renderer.cacheMap) {
            if (!renderer.cacheMap.hasOwnProperty(declarationReference)) {
              const pluginInterface = {
                property,
                value,
                pseudo,
                media,
                support,
              }

              const processed = arrayReduce(
                renderer.optimizedPlugins,
                applyPlugin,
                pluginInterface
              )

              const cacheReference = generateDeclarationReference(
                processed.property,
                processed.value,
                processed.pseudo,
                processed.media,
                processed.support
              )

              if (!renderer.cache.hasOwnProperty(cacheReference)) {
                renderer._renderStyleToCache(
                  cacheReference,
                  processed.property,
                  processed.value,
                  processed.pseudo,
                  processed.media,
                  processed.support
                )
              }

              renderer.cacheMap[declarationReference] = cacheReference
            }

            declarationReference = renderer.cacheMap[declarationReference]
          }

          if (!renderer.cache.hasOwnProperty(declarationReference)) {
            renderer._renderStyleToCache(
              declarationReference,
              property,
              value,
              pseudo,
              media,
              support
            )
          }

          const cachedClassName = renderer.cache[declarationReference].className

          // only append if we got a class cached
          if (cachedClassName) {
            classNames += ' ' + cachedClassName
          }
        }
      }

      return classNames
    },

    _renderStyleToCache(reference, property, value, pseudo, media, support) {
      // we remove undefined values to enable
      // usage of optional props without side-effects
      if (isUndefinedValue(value)) {
        renderer.cache[reference] = {
          className: '',
        }

        return
      }

      const className =
        renderer.selectorPrefix +
        renderer.generateClassName(property, value, pseudo, media, support)

      const declaration = cssifyDeclaration(property, value)
      const selector = generateCSSSelector(
        className,
        pseudo,
        config.specificityPrefix,
        renderer.propertyPriority[property]
      )

      const change = {
        type: RULE_TYPE,
        className,
        selector,
        declaration,
        pseudo,
        media,
        support,
      }

      renderer.cache[reference] = change
      renderer._emitChange(change)
    },

    generateClassName() {
      return generateClassName(
        renderer.getNextRuleIdentifier,
        renderer.filterClassName
      )
    },

    _emitChange(change) {
      arrayEach(renderer.listeners, (listener) => listener(change))
    },
  }

  // initial setup
  renderer.keyframePrefixes.push('')

  renderer.optimizedPlugins = arrayReduce(
    renderer.plugins,
    (plugins, plugin) => {
      if (plugin.optimized) {
        plugins.push(plugin.optimized)
      }

      return plugins
    },
    []
  )

  // only enable the cache map if we have optimized plugins
  if (renderer.optimizedPlugins.length > 0) {
    renderer.cacheMap = {}
    renderer.unoptimizedPlugins = arrayFilter(
      renderer.plugins,
      (plugin) => !plugin.optimized
    )
  }

  if (config.enhancers) {
    arrayEach(config.enhancers, (enhancer) => {
      renderer = enhancer(renderer)
    })
  }

  return renderer
}
