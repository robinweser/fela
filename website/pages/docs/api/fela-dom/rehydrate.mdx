import DocLayout from '../../../../components/DocLayout.js'

export default DocLayout

# rehydrate

> This method is used for universal rendering where styles are prerender on the server. For client-side only rendering, we recommend using [render](render.md) directly.

Rehydrates the renderer's cache from existing `<style>` elements.<br />It is primarily used to rehydrate the cache from server-rendered CSS.

It automatically connects the renderer with the DOM by calling [render](render.md) once finished rehydrating.

## Arguments

| Argument | Type                              | Description                               |
| -------- | --------------------------------- | ----------------------------------------- |
| renderer | [_Renderer_](../fela/Renderer.md) | The renderer which's cache is rehydrated. |

## Example

Assuming we have the following server-rendered elements:

```html
<style data-fela-type="RULE">
  .a {
    color: red;
  }
  .b {
    color: blue;
  }
</style>
```

```javascript
import { createRenderer } from 'fela'
import { rehydrate } from 'fela-dom'

const renderer = createRenderer()

rehydrate(renderer)

const rule = (props) => ({
  color: props.color,
})

renderer.renderRule(rule, { color: 'red' }) // => a
renderer.renderRule(rule, { color: 'green' }) // => c
renderer.renderRule(rule, { color: 'blue' }) // => b
```
