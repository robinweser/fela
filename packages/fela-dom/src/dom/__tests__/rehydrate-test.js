import rehydrate from '../rehydrate'

import renderToMarkup from '../../server/renderToMarkup'
import createRenderer from '../../../../fela/src/createRenderer'
import webPreset from '../../../../fela-preset-web/src/index'

import sortObject from '../__helpers__/sortObject'

beforeEach(() => {
  const { head } = document
  while (head.firstChild) {
    head.removeChild(head.firstChild)
  }
})

describe('Rehydrating from DOM nodes', () => {
  it('should rehydrate the renderer cache', () => {
    const serverRenderer = createRenderer({
      filterClassName: cls => cls !== 'a',
      plugins: [...webPreset],
    })

    serverRenderer.renderRule(() => ({
      color: 'yellow',
      backgroundColor: 'red',
      flex: 1,
      '& #id > .foo ~ bar': {
        backgroundColor: 'red',
      },
      '[alt="Hello"]': {
        fontSize: 12,
      },
      '@supports (display: grid)': {
        color: 'blue',
        '&.foo.bar': {
          color: 'red',
        },
      },
      ':hover': {
        color: 'red',
        '> h1': {
          color: 'green',
        },
      },
    }))

    document.head.innerHTML = renderToMarkup(serverRenderer)

    const clientRenderer = createRenderer({
      filterClassName: cls => cls !== 'a',
      plugins: [...webPreset],
    })

    rehydrate(clientRenderer)

    expect([
      clientRenderer.uniqueRuleIdentifier,
      clientRenderer.cache,
    ]).toMatchSnapshot()

    expect(sortObject(clientRenderer.cache)).toEqual(
      sortObject(serverRenderer.cache)
    )
  })

  it('should continue in rehydration with correct rehydration index', () => {
    const serverRenderer = createRenderer({
      plugins: [...webPreset],
    })

    serverRenderer.renderRule(() => ({
      color: 'yellow',
      backgroundColor: 'red',
    }))

    expect(serverRenderer.uniqueRuleIdentifier).toBe(2)

    document.head.innerHTML = renderToMarkup(serverRenderer)

    const clientRenderer = createRenderer({
      plugins: [...webPreset],
    })

    rehydrate(clientRenderer)

    expect(clientRenderer.uniqueRuleIdentifier).toBe(2)

    // simulates e.g. page transition with new component rule
    rehydrate(clientRenderer)

    clientRenderer.renderRule(() => ({
      color: 'black',
    }))

    expect(clientRenderer.uniqueRuleIdentifier).toBe(3)

    // simulates e.g. page transition with new component rule
    rehydrate(clientRenderer)

    clientRenderer.renderRule(() => ({
      color: 'purple',
    }))

    expect(clientRenderer.uniqueRuleIdentifier).toBe(4)
  })
})
