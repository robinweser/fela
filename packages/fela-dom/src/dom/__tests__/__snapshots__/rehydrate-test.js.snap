// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rehydrating from DOM nodes should rehydrate the renderer cache 1`] = `
[
  10,
  {
    " #id > .foo ~ barWebkitBackgroundColorred;background-color:red": {
      "className": "e",
      "declaration": "-webkit-background-color:red;background-color:red",
      "media": "",
      "pseudo": " #id > .foo ~ bar",
      "selector": ".e #id > .foo ~ bar",
      "support": "",
      "type": "RULE",
    },
    "(display: grid) .foo.barcolorred": {
      "className": "h",
      "declaration": "color:red",
      "media": "",
      "pseudo": " .foo.bar",
      "selector": ".h .foo.bar",
      "support": "(display: grid)",
      "type": "RULE",
    },
    "(display: grid)colorblue": {
      "className": "g",
      "declaration": "color:blue",
      "media": "",
      "pseudo": "",
      "selector": ".g",
      "support": "(display: grid)",
      "type": "RULE",
    },
    ":hover> h1colorgreen": {
      "className": "j",
      "declaration": "color:green",
      "media": "",
      "pseudo": ":hover> h1",
      "selector": ".j:hover> h1",
      "support": "",
      "type": "RULE",
    },
    ":hovercolorred": {
      "className": "i",
      "declaration": "color:red",
      "media": "",
      "pseudo": ":hover",
      "selector": ".i:hover",
      "support": "",
      "type": "RULE",
    },
    "WebkitBackgroundColorred;background-color:red": {
      "className": "c",
      "declaration": "-webkit-background-color:red;background-color:red",
      "media": "",
      "pseudo": "",
      "selector": ".c",
      "support": "",
      "type": "RULE",
    },
    "WebkitFlex1;-ms-flex:1;flex:1": {
      "className": "d",
      "declaration": "-webkit-flex:1;-ms-flex:1;flex:1",
      "media": "",
      "pseudo": "",
      "selector": ".d",
      "support": "",
      "type": "RULE",
    },
    "[alt=\"Hello\"]fontSize12px": {
      "className": "f",
      "declaration": "font-size:12px",
      "media": "",
      "pseudo": "[alt=\"Hello\"]",
      "selector": ".f[alt=\"Hello\"]",
      "support": "",
      "type": "RULE",
    },
    "coloryellow": {
      "className": "b",
      "declaration": "color:yellow",
      "media": "",
      "pseudo": "",
      "selector": ".b",
      "support": "",
      "type": "RULE",
    },
  },
]
`;

exports[`Rehydrating from DOM nodes should rehydrate the renderer cache 2`] = `
[
  10,
  {
    " #id > .foo ~ barWebkitBackgroundColorred;background-color:red": {
      "className": "e",
      "declaration": "-webkit-background-color:red;background-color:red",
      "media": "",
      "pseudo": " #id > .foo ~ bar",
      "selector": ".parentClass .e #id > .foo ~ bar",
      "support": "",
      "type": "RULE",
    },
    "(display: grid) .foo.barcolorred": {
      "className": "h",
      "declaration": "color:red",
      "media": "",
      "pseudo": " .foo.bar",
      "selector": ".parentClass .h .foo.bar",
      "support": "(display: grid)",
      "type": "RULE",
    },
    "(display: grid)colorblue": {
      "className": "g",
      "declaration": "color:blue",
      "media": "",
      "pseudo": "",
      "selector": ".parentClass .g",
      "support": "(display: grid)",
      "type": "RULE",
    },
    ":hover> h1colorgreen": {
      "className": "j",
      "declaration": "color:green",
      "media": "",
      "pseudo": ":hover> h1",
      "selector": ".parentClass .j:hover> h1",
      "support": "",
      "type": "RULE",
    },
    ":hovercolorred": {
      "className": "i",
      "declaration": "color:red",
      "media": "",
      "pseudo": ":hover",
      "selector": ".parentClass .i:hover",
      "support": "",
      "type": "RULE",
    },
    "WebkitBackgroundColorred;background-color:red": {
      "className": "c",
      "declaration": "-webkit-background-color:red;background-color:red",
      "media": "",
      "pseudo": "",
      "selector": ".parentClass .c",
      "support": "",
      "type": "RULE",
    },
    "WebkitFlex1;-ms-flex:1;flex:1": {
      "className": "d",
      "declaration": "-webkit-flex:1;-ms-flex:1;flex:1",
      "media": "",
      "pseudo": "",
      "selector": ".parentClass .d",
      "support": "",
      "type": "RULE",
    },
    "[alt=\"Hello\"]fontSize12px": {
      "className": "f",
      "declaration": "font-size:12px",
      "media": "",
      "pseudo": "[alt=\"Hello\"]",
      "selector": ".parentClass .f[alt=\"Hello\"]",
      "support": "",
      "type": "RULE",
    },
    "coloryellow": {
      "className": "b",
      "declaration": "color:yellow",
      "media": "",
      "pseudo": "",
      "selector": ".parentClass .b",
      "support": "",
      "type": "RULE",
    },
  },
]
`;
